dev/com.ibm.ws.jsp/src/com/ibm/ws/jsp/translator/visitor/generator/GenerateJspVisitor.java

```diff
@@ -191,73 +195,95 @@ else if (methodWriter instanceof CleanupTaglibLookupWriter) {
 				writer.println("deferredSyntaxAllowedAsLiteral = [" + jspConfiguration.isDeferredSyntaxAllowedAsLiteral() +"]"); // jsp2.1ELwork
 
 				writer.println("*/");
-				
-				
-				
+
+
+
 			}
 
         }
 
     }
 
+    // Added for Pages 3.1's errorOnELNotFound option
+    private void generateIsErrorOnELFoundMethod(boolean flag) {
+
+        writer.println(" public boolean isErrorOnELNotFound() {");
+        writer.println("return "+ flag  + ";");
+        writer.println("}");
+    }
+
     protected void generateImportSection(ValidateJspResult validatorResult) {
         String servletPackageName = jspPackageName;
 
         writer.println("package " + servletPackageName + ";");
         writer.println();
-        
+
         for(int i = 0; i < Constants.STANDARD_IMPORTS.length; i++)
             writer.println("import " + (String) Constants.STANDARD_IMPORTS[i] + ";");
-        
+
         writer.println();
     }
-    
+
     protected void generateClassSection(ValidateJspResult validatorResult) {
         writer.println();
         writer.print("public final class " + jspClassName + " extends ");
         String extendsClass = validatorResult.getExtendsClass();
         writer.print(extendsClass == null
                 ? jspServletBase
                 : extendsClass);
-        interfaces.add("com.ibm.ws.jsp.runtime.JspClassInformation");                
-        boolean singleThreaded = validatorResult.isSingleThreaded();
-        if (singleThreaded)
-            interfaces.add("SingleThreadModel");
-        
+        interfaces.add("com.ibm.ws.jsp.runtime.JspClassInformation");
+        if(PagesVersionHandler.isPages31OrHigherLoaded()){
+            interfaces.add("com.ibm.ws.jsp.runtime.JspDirectiveInfo");
+        }
+
+        /*  SingleThreadModel was removed in Servlet 6.0 (Pages 3.1)
+         *  For Page 3.1, the jsp service method is synchronized instead (performance hit)
+         *  isThreadSafe is not recommended anymore. 
+         */ 
+        if(PagesVersionHandler.isPages30OrLowerLoaded()){
+            boolean singleThreaded = validatorResult.isSingleThreaded();
+
+            if (singleThreaded)
+                interfaces.add("SingleThreadModel");
+        } else {
+            logger.logp(Level.WARNING, CLASS_NAME, "generateClassSection", "jsp.isthreadsafe.warning");
+        }
+
```

Comment:
I think the logic here is bad. Needs to be re-done.

Reply:
Since we are in common code here we should clean it up before merging IMO. What part of the logic specifically are you worried about?

Reply:
I think I need to add a singleThreaded check before logging the warning.  It looks like it would be logged every time the pages 3.1 feature is used.